ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -209,25 +285,108 @@ boolean isEncryptionEnabled() {
-      public void onCompletion(Long result, Exception exception) {
-        if (exception != null) {
-          setOperationExceptionAndComplete(exception);
-          routerCallback.onPollReady();
-        } else {
-          blobSize = result;
-          chunkFillingCompletedSuccessfully = true;
-        }
-        chunkFillerChannel.close();
+  private void startReadingFromChannel() {
+    channel.readInto(chunkFillerChannel, (result, exception) -> {
+      if (exception != null) {
+        setOperationExceptionAndComplete(exception);
+        routerCallback.onPollReady();
+      } else {
+        blobSize = result;
+        chunkFillingCompletedSuccessfully = true;
       }
+      chunkFillerChannel.close();
     });
   }
 
+  /**
+   * Process the list of data chunks provided and prepare a metadata chunk with the provided chunk IDs.
+   * @throws RouterException if validation of the chunks to stitch failed.
+   */
+  private void processChunksToStitch() throws RouterException {
+    if (chunksToStitch.isEmpty()) {
+      // The current metadata format does not support empty chunk lists
+      throw new RouterException("Must provide at least one chunk for stitchBlob call",
+          RouterErrorCode.InvalidPutArgument);
+    }
+    long totalSize = 0;
+    long intermediateChunkSize = chunksToStitch.get(0).getChunkSizeInBytes();
+    metadataPutChunk.setIntermediateChunkSize(intermediateChunkSize);
+    for (ListIterator<ChunkInfo> iter = chunksToStitch.listIterator(); iter.hasNext(); ) {
+      int chunkIndex = iter.nextIndex();
+      ChunkInfo chunkInfo = iter.next();
+      BlobId chunkId = unwrapChunkInfo(chunkInfo, intermediateChunkSize, !iter.hasNext());
+      metadataPutChunk.addChunkId(chunkId, chunkIndex);
+      totalSize += chunkInfo.getChunkSizeInBytes();
+    }
+    blobSize = totalSize;
+    chunkFillingCompletedSuccessfully = true;
+  }
+
+  /**
+   * Extract the {@link BlobId} from a {@link ChunkInfo} object and check the metadata in the object to ensure that this
+   * is a valid chunk for stitching.
+   * @param chunkInfo the {@link ChunkInfo} containing the blob ID and extra metadata.
+   * @param intermediateChunkSize the size in bytes that intermediate chunks must be with the current metadata format.
+   * @param lastChunk {@code true} if this is the last chunk in the list of blobs to stitch.
+   * @return the {@link BlobId}.
+   * @throws RouterException if validation of the chunk or blob ID parsing failed.
+   */
+  private BlobId unwrapChunkInfo(ChunkInfo chunkInfo, long intermediateChunkSize, boolean lastChunk)
+      throws RouterException {
+    long chunkSize = chunkInfo.getChunkSizeInBytes();
+    long chunkExpirationTimeInMs = chunkInfo.getExpirationTimeInMs();
+
+    if (chunkSize == 0 || (lastChunk ? chunkSize > intermediateChunkSize : chunkSize != intermediateChunkSize)) {
+      throw new RouterException("Invalid chunkSize: " + chunkSize + "; intermediateChunkSize: " + intermediateChunkSize,
```

Comment:
the message could also include if this was the last chunk.

Reply:
Also, I find the following more readable:
```java
if (chunkSize == 0 || chunkSize > intermediateChunkSize || (!lastChunk && chunkSize < intermediateChunkSize)) {
...
}
```

Reply:
Thanks! that is more readable!